### PR TITLE
Putting self-contained icon upload behind feature flag

### PR DESCRIPTION
**Pull request by agr**
Addresses https://github.com/NuGet/NuGetGallery/issues/7090

Added a `NuGetGallery.EmbeddedIcons` feature flag that blocks upload of packages with embedded icons except for a chosen few. Had to pass `currentUser` into `PackageUploadService`, so this involves quite a bit of changes here and there.
Added tests for rejection/acceptance cases.
Fixed the brittleness of the `RejectsNupkgsReportingIncorrectFileLengthForNuspecFile` test introduced in #7038. Previously it relied on a fact that the `theId.nuspec` file included as a resource was the same as the nuspec file generated by `GeneratePackageStream()` call, which broke when the default file content changed. Replaced that with getting the actual nuspec file from the nupkg.

### CODE REVIEW

File: src/NuGetGallery/Services/PackageUploadService.cs [hunk 1]
```diff
@@ -87,9 +90,13 @@ public class PackageUploadService : IPackageUploadService
...
             }
             _trace = diagnosticsService.GetSource(nameof(PackageUploadService));
+            _featureFlagService = featureFlagService ?? throw new ArgumentNullException(nameof(featureFlagService));
         }
 
-        public async Task<PackageValidationResult> ValidateBeforeGeneratePackageAsync(PackageArchiveReader nuGetPackage, PackageMetadata packageMetadata)
+        public async Task<PackageValidationResult> ValidateBeforeGeneratePackageAsync(
+            PackageArchiveReader nuGetPackage,
+            PackageMetadata packageMetadata,
+            User user)
```
Comment (scottbommarito) [inline src/NuGetGallery/Services/PackageUploadService.cs]: Maybe consider a more descriptive name here? `currentUser` or `uploader`?

File: src/NuGetGallery/Services/PackageUploadService.cs [hunk 2]
```diff
@@ -135,13 +142,9 @@ public async Task<PackageValidationResult> ValidateBeforeGeneratePackageAsync(Pa
             return PackageValidationResult.AcceptedWithWarnings(warnings);
         }
 
-        private async Task<PackageValidationResult> CheckLicenseMetadataAsync(PackageArchiveReader nuGetPackage, List<IValidationMessage> warnings)
+        private async Task<PackageValidationResult> CheckLicenseMetadataAsync(PackageArchiveReader nuGetPackage, List<IValidationMessage> warnings, User user)
```
Comment (scottbommarito) [inline src/NuGetGallery/Services/PackageUploadService.cs]: This method checks icon validity now too. Should we rename it to `CheckUserContentAsync` or something?

File: tests/NuGetGallery.Core.Facts/TestUtils/TestPackage.cs [hunk 3]
```diff
@@ -229,6 +235,22 @@ private static string WriteDependencies(IEnumerable<PackageDependencyGroup> pack
             }, desiredTotalEntryCount);
         }
 
+        private static string AddBinaryFile(ZipArchive archive, string filename, byte[] fileContents)
+        {
+            if (fileContents != null && filename != null)
+            {
+                // enforce directory separators the same way as the client (see PackageArchiveReader.GetStream)
+                filename = PathUtility.StripLeadingDirectorySeparators(filename);
+                var licenseEntry = archive.CreateEntry(filename, CompressionLevel.Fastest);
```
Comment (scottbommarito) [inline tests/NuGetGallery.Core.Facts/TestUtils/TestPackage.cs]: update this variable name

File: tests/NuGetGallery.Core.Facts/TestUtils/TestPackage.cs [hunk 4]
```diff
@@ -229,6 +235,22 @@ private static string WriteDependencies(IEnumerable<PackageDependencyGroup> pack
...
+                using (var licenseStream = licenseEntry.Open())
+                {
+                    licenseStream.Write(fileContents, 0, fileContents.Length);
```
Comment (scottbommarito) [inline tests/NuGetGallery.Core.Facts/TestUtils/TestPackage.cs]: update this variable name

File: src/NuGetGallery/Services/IFeatureFlagService.cs [hunk 5]
```diff
@@ -42,5 +42,10 @@ public interface IFeatureFlagService
         /// </summary>
         /// <returns></returns>
         bool IsSearchCircuitBreakerEnabled();
+
+        /// <summary>
+        /// Whether or not the packages with embedded icons are allowed for publishing.
+        /// </summary>
+        bool IsEmbeddedIconsEnabled(User user);
```
Comment (scottbommarito) [inline src/NuGetGallery/Services/IFeatureFlagService.cs]: `IsEmbeddedIconsEnabled` -> `AreEmbeddedIconsEnabled`

File: src/NuGetGallery/Services/PackageUploadService.cs (hunk 2 above)
Comment (agr) [reply src/NuGetGallery/Services/PackageUploadService.cs]: I will extract icon related stuff into separate method for proper icon validation (or move all validation stuff into class of its own, this gets large). For now, I didn't want to have nuspec parsed twice between license file and icon validations.

File: src/NuGetGallery/Services/PackageUploadService.cs [hunk 6]
```diff
@@ -184,6 +187,12 @@ private async Task<PackageValidationResult> CheckLicenseMetadataAsync(PackageArc
             var licenseMetadata = nuspecReader.GetLicenseMetadata();
             var licenseDeprecationUrl = GetExpectedLicenseUrl(licenseMetadata);
 
+            // TODO: move out when full blown validation is implemented
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/PackageUploadService.cs]: >TODO [](start = 15, length = 4)

Please add a link to a GitHub issue so we can track this

File: src/NuGetGallery/Services/IFeatureFlagService.cs [hunk 7]
```diff
@@ -42,5 +42,10 @@ public interface IFeatureFlagService
         /// </summary>
         /// <returns></returns>
         bool IsSearchCircuitBreakerEnabled();
+
+        /// <summary>
+        /// Whether or not the packages with embedded icons are allowed for publishing.
+        /// </summary>
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/IFeatureFlagService.cs]: What do you think of something like: `Whether the user is allowed to publish packages with an embedded icon.`?

**Pull request merged**